### PR TITLE
Parallelize validate

### PR DESCRIPTION
**Pull request by ssssarah**
- the traverse function being optionally overwritten on Shape was due to a change in pyshacl versions (I believe an old version had this version but a newer one no longer had it. For the version we have fixed, this function does not exist) https://github.com/BlueBrain/nexus-forge/issues/43 https://github.com/BlueBrain/nexus-forge/pull/44 
- renamed usage of `_transitive_load_shape_graph` to `_transitive_load_resource_graph`
- dummy implementation of `_validate_many` in `DemoModel`
- Override pyshacl `ShapesGraph` implementation of `def shapes` to return `list` instead of `dict_values()` so that it can be picklable (for multiprocessing)
- Parallelized validate. 
  -> cannot use threads because pyshacl uses rdflib, rdflib uses pyparsing and it is not thread safe
  -> shape loading occurs before creating multiple processes for validation. It would be possible to load shapes in individual processes, however: 
       - "caching" of shapes becomes useless, the same shape will be queried multiple times across processes 
       - Shape loading in individual processes would be best if validating resources of different types/shapes
       - The former point might not be completely true because shapes re-use each other
       - It is for sure better to load the shapes beforehand if we're dealing with a list of resources of the same type/shape.
    Given these reasons, I opted for loading all shapes beforehand.

### CODE REVIEW

**Comment by ssssarah** [comment]
Unfortunately I ran into an issue when testing validation on 4000 resources. 
Long story short, pyshacl uses rdflib, rdflib uses pyparsing and it is not thread safe 
https://github.com/RDFLib/rdflib/issues/765 I'll look more into it

File: kgforge/specializations/models/rdf_model.py [hunk 1]
```diff
@@ -127,33 +130,72 @@ def validate(
...
+
+        resource._validated = conforms
+        resource._last_action = Action(operation="_validate_many", succeeded=conforms, error=err)
+
+        if not conforms and raise_:
+            raise err
+
+        return resource
+
+    def _get_shape_stuff(self, r: Resource, type_: str) -> Tuple[str, ShapeWrapper, Graph, Graph, Resource]:
```
Comment (crisely09) [inline kgforge/specializations/models/rdf_model.py:163]: not convinced about this name

File: kgforge/specializations/models/rdf/service.py [hunk 2]
```diff
@@ -138,12 +137,19 @@ def lookup_shape_from_node(self, node: URIRef) -> Shape:
...
-            shape_wrapper = ShapeWrapper(self._node_shape_cache[node])
-            if not hasattr(shape_wrapper, "traverse"):
-                shape_wrapper.traverse = types.MethodType(traverse, shape_wrapper)
-            return shape_wrapper
-        return shape
+
+        return ShapeWrapper(shape)
+
+    @property
+    def shapes(self):  # pyshacl implementation returns dict_values (not list). This cannot be pickled.
```
Comment (crisely09) [inline kgforge/specializations/models/rdf/service.py:144]: maybe `shapes_list` would be even more accurate?

File: kgforge/specializations/models/rdf_model.py (hunk 1 above)
Comment (ssssarah) [reply kgforge/specializations/models/rdf_model.py:163]: that is fair 😅

File: kgforge/specializations/models/rdf/service.py (hunk 2 above)
Comment (ssssarah) [reply kgforge/specializations/models/rdf/service.py:144]: This is overriding a property defined by pyshacl, so I have to stick to the same name

Comment (crisely09) [reply kgforge/specializations/models/rdf/service.py:144]: Ah ok

**Review by crisely09 — APPROVED** [review]

**Pull request merged**